account_fiscal_product_rule/tests/test_account_fiscal_product_rule.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Akretion France (http://www.akretion.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo.tests.common import SavepointCase
+
+
+class TestAccountFiscalProductRule(SavepointCase):
```

Comment:
In order to reduce the setup, it will be better to re-use the existing setup class of odoo.
https://github.com/odoo/odoo/blob/657fd6653705255f319756cba6406cefed9649d4/addons/account/tests/common.py#L11

You can do 

from odoo.addons.account.tests.common import AccountTestInvoicingCommon


Then for the test you can use

self.product_a
self.partner_b (that have the fiscal position a) : https://github.com/odoo/odoo/blob/657fd6653705255f319756cba6406cefed9649d4/addons/account/tests/common.py#L148
and you can use the method : init_invoice (with the partner_b and the product a : https://github.com/odoo/odoo/blob/657fd6653705255f319756cba6406cefed9649d4/addons/account/tests/common.py#L366)

Then in your setup you only need to create
- a new tax (you can copy it, with a new rate copy(amount=30))
- a new receivable account (use the method copy_account : https://github.com/odoo/odoo/blob/657fd6653705255f319756cba6406cefed9649d4/addons/account/tests/common.py#L14)
- a fiscal product rule


For the test the aim is to test the impact of the fiscal_product_rule

So you should write 3 tests

def test_no_rule(self):
    invoice = self.init_invoice(partner=self.partner_b, products=self.product_a)
    # check the tax/account

def test_rule_on_categ(self):
    self.product.categ_id.fiscal_position_product_rule_ids = self.fiscal_product_rule
    invoice = self.init_invoice(partner=self.partner_b, products=self.product_a)
    # check the tax/account is the on define by the rule

def test_rule_on_product(self):
    self.product.fiscal_position_product_rule_ids = self.fiscal_product_rule
    invoice = self.init_invoice(partner=self.partner_b, products=self.product_a)
    # check the tax/account is the on define by the rule


With that solution you should reduce a lot the number of line of your test

Reply:
@sebastienbeau , modifications made on the tests following your comments.